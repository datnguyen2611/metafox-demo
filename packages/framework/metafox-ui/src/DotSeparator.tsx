import { Box, SxProps } from '@mui/material';
import { styled } from '@mui/material/styles';
import React from 'react';

const StyledSeparator = styled(Box, {
  name: 'DotSeparator',
  shouldForwardProp: (prop: string) =>
    !/isTruncateOneLine|placement|separator/.test(prop)
})<{
  placement?: 'start' | 'end';
  isTruncateOneLine?: boolean;
  separator: string;
}>(({ placement, separator, isTruncateOneLine }) => ({
  // use when CategoryList wants to have 1 line and truncate the category
  ...(isTruncateOneLine && {
    display: 'block !important',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    maxWidth: '100%'
  }),
  ...(placement === 'end' && {
    '&>span:not(:last-child):after': {
      textDecoration: 'none',
      content: separator ?? '"·"',
      paddingLeft: '0.4em',
      paddingRight: '0.4em',
      fontWeight: 'normal'
    }
  }),
  ...(placement === 'start' && {
    '&>span:not(:first-of-type):before': {
      textDecoration: 'none',
      content: separator ?? '"·"',
      paddingLeft: '0.4em',
      paddingRight: '0.4em',
      fontWeight: 'normal'
    }
  })
}));

export default function DotSeparator({
  placement = 'end',
  isTruncateOneLine = false,
  children,
  separator = '"·"',
  ...props
}: {
  className?: string;
  placement?: 'end' | 'start';
  isTruncateOneLine?: boolean;
  separator?: string;
  sx?: SxProps;
  children?: React.ReactNode;
}) {
  return (
    <StyledSeparator
      separator={separator}
      isTruncateOneLine={isTruncateOneLine}
      placement={placement}
      {...props}
    >
      {React.Children.toArray(children).map((node, index) => (
        <span key={index.toString()} children={node} />
      ))}
    </StyledSeparator>
  );
}
