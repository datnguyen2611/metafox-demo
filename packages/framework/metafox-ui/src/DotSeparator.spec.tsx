import { render } from '@testing-library/react';
import * as React from 'react';
import DotSeparator from './DotSeparator';

const MockChildComponent = ({ index }) => <>item {index}</>;

describe('DotSeparator', () => {
  const props = {
    sx: {
      color: 'text.secondary',
      mt: 1
    }
  };

  it('render empty item', async () => {
    const { container } = render(<DotSeparator {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('render have item', async () => {
    const { container } = render(
      <DotSeparator {...props}>
        <MockChildComponent index={1} />
        <MockChildComponent index={2} />
      </DotSeparator>
    );

    expect(container).toMatchSnapshot();
  });
});
