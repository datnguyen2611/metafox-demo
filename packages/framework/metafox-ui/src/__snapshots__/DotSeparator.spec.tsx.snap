// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DotSeparator render empty item 1`] = `
<div>
  <div
    class="MuiBox-root css-1fhe8qf-DotSeparator-root"
  />
</div>
`;

exports[`DotSeparator render have item 1`] = `
<div>
  <div
    class="MuiBox-root css-1fhe8qf-DotSeparator-root"
  >
    <span>
      item 
      1
    </span>
    <span>
      item 
      2
    </span>
  </div>
</div>
`;
